src/celengine/overlay.cpp

```diff
@@ -161,27 +166,106 @@ void Overlay::print(const char* s)
     }
 }
 
-void Overlay::rect(float x, float y, float w, float h, bool fill)
+static void drawRect(const Renderer& r,
+                     const array<float, 8>& vertices,
+                     const vector<Vector4f>& colors,
+                     Overlay::RectType type,
+                     float linewidth)
 {
-    if (useTexture)
+
+    CelestiaGLProgram *prog;
+
+    ShaderProperties shader;
+    shader.simpleProps = 0;
+
+    switch (type)
     {
-        glDisable(GL_TEXTURE_2D);
-        useTexture = false;
+    case Overlay::RectType::Textured:
+        shader.simpleProps |= ShaderProperties::HasTexture;
+    // [[ fallthrough ]]
+    case Overlay::RectType::Outlined:
+    case Overlay::RectType::Filled:
+        switch (colors.size())
+        {
+        case 0:
+            break;
+        case 1:
+            shader.simpleProps |= ShaderProperties::UniformColor;
+            break;
+        case 4:
+            shader.simpleProps |= ShaderProperties::PerVertexColor;
+            break;
+        default:
+            fmt::fprintf(cerr, "Incorrect number of colors: %i\n", colors.size());
+            break;
+        }
+    default:
+        break;
     }
+    prog = r.getShaderManager().getShader(shader);
+
+    if (prog == nullptr)
+        return;
 
-    if (fill)
+    constexpr array<short, 8> texels = {0, 1,  1, 1,  1, 0, 0, 0};
```

Comment:
Double spaces in two cases.

Reply:
actually one space is missing.

these are texture coordinates x0, y0, x1, y1, x2, y2, x3, y3 so with additional spaces it's seen easier where one texel finishes and another one starts.